### PR TITLE
feat(engine): invalid block hooks crate

### PR DESCRIPTION
**Pull request by shekhirin**
Closes https://github.com/paradigmxyz/reth/issues/10546

### CODE REVIEW

File: crates/engine/tree/src/tree/invalid_block_hook.rs [hunk 1]
```diff
@@ -49,3 +49,31 @@ impl InvalidBlockHook for NoopInvalidBlockHook {
     ) {
     }
 }
+
+/// A chain of [`InvalidBlockHook`]s that are executed in order.
+pub struct InvalidBlockHookChain(pub Vec<Box<dyn InvalidBlockHook>>);
```
Comment (mattsse) [inline crates/engine/tree/src/tree/invalid_block_hook.rs:54]: can we replace `Chain` with something else, maybe even just `InvalidBlockHooks`

File: crates/engine/tree/src/tree/invalid_block_hook.rs [hunk 2]
```diff
@@ -49,3 +49,31 @@ impl InvalidBlockHook for NoopInvalidBlockHook {
...
+        f.debug_struct("InvalidBlockHookChain").field("len", &self.0.len()).finish()
+    }
+}
+
+impl InvalidBlockHook for InvalidBlockHookChain {
+    fn on_invalid_block(
+        &self,
+        block: SealedBlockWithSenders,
+        header: SealedHeader,
+        output: BlockExecutionOutput<Receipt>,
```
Comment (mattsse) [inline crates/engine/tree/src/tree/invalid_block_hook.rs:67]: should we make all of these & and expect the hooks to clone on demand?

File: crates/node/builder/src/launch/engine.rs [hunk 3]
```diff
@@ -202,12 +202,28 @@ where
...
+                            Ok(Box::new(reth_invalid_block_hooks::witness)
+                                as Box<dyn InvalidBlockHook>)
+                        }
+                        reth_node_core::args::InvalidBlockHook::PreState |
+                        reth_node_core::args::InvalidBlockHook::Opcode => {
+                            eyre::bail!("invalid block hook {hook:?} is not implemented yet")
+                        }
+                    })
+                    .collect::<Result<Vec<_>, _>>()?,
+            ))
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/node/builder/src/launch/engine.rs:223]: this should be moved into the LaunchCtx type

File: crates/node/builder/src/launch/engine.rs [hunk 4]
```diff
@@ -202,12 +202,28 @@ where
...
-        }
-
-        let invalid_block_hook = Box::new(NoopInvalidBlockHook);
+        let invalid_block_hook: Box<dyn InvalidBlockHook> = if let Some(ref hook) =
+            ctx.node_config().debug.invalid_block_hook
+        {
+            warn!(target: "reth::cli", ?hook, "Invalid block hooks are not implemented yet! The `debug.invalid-block-hook` flag will do nothing for now.");
+            Box::new(InvalidBlockHookChain(
+                hook.to_selection()
+                    .into_iter()
```
Comment (mattsse) [inline crates/node/builder/src/launch/engine.rs:211]: isn't this just iter().copied() ?

**Review by mattsse — CHANGES_REQUESTED** [review]
some suggestions

File: crates/node/builder/src/launch/engine.rs (hunk 4 above)
Comment (shekhirin) [reply crates/node/builder/src/launch/engine.rs]: yep, added a `Deref` impl to the selection struct and replace with `iter().copied()` here

File: crates/engine/tree/src/tree/invalid_block_hook.rs (hunk 2 above)
Comment (shekhirin) [reply crates/engine/tree/src/tree/invalid_block_hook.rs]: agreed

File: crates/engine/tree/src/tree/invalid_block_hook.rs (hunk 1 above)
Comment (shekhirin) [reply crates/engine/tree/src/tree/invalid_block_hook.rs]: no chain in blockchain project

image

Comment (shekhirin) [reply crates/engine/tree/src/tree/invalid_block_hook.rs]: but agreed, just `InvalidBlockHooks` is easier

File: crates/node/builder/src/launch/common.rs [hunk 5]
```diff
@@ -188,7 +189,7 @@ impl LaunchContext {
 /// This can be used to sequentially attach additional values to the type during the launch process.
 ///
 /// The type provides common boilerplate for launching a node depending on the additional value.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
```
Comment (mattsse) [inline crates/node/builder/src/launch/common.rs:192]: why is this no longer clone?

Comment (mattsse) [inline crates/node/builder/src/launch/common.rs:192]:
```suggestion
#[derive(Debug, Clone)]
```

**Review by mattsse — APPROVED** [review]

**Pull request merged**